### PR TITLE
Obligations shrinking: shrink abstraction too

### PR DESCRIPTION
**Pull request by mattam82**
This is a fix to the obligation shrinking option which was repeating lets in the abstraction even when thoses were not used by the obligation, defeating the shrinking phase and resulting in very poor behavior when evaluating programs in call-by-value fashion. It changes the terms produced by Program for the better.

### CODE REVIEW

File: vernac/obligations.ml [hunk 1]
```diff
@@ -631,6 +631,16 @@ let unfold_entry cst = Hints.HintsUnfoldEntry [EvalConstRef cst]
 let add_hint local prg cst =
   Hints.add_hints local [Id.to_string prg.prg_name] (unfold_entry cst)
 
+let it_mkLambda_or_LetIn_or_clean t ctx =
```
Comment (maximedenes) [inline vernac/obligations.ml]: Could you add a header comment saying what this function does? I'm not sure to fully understand why we do `if noccurn 1 t then subst1 mkProp t`.

Comment (maximedenes) [inline vernac/obligations.ml]: Also, is it really the right place for such a function?

**Comment by herbelin** [comment]
That's right. This function could be in `Termops` which makes me see btw that `Econstr.it_mkLambda_or_LetIn` seems redundant with `Termops.it_mkLambda_or_LetIn`.

However, "clean" did not ring a clear bell to me about what it did. Don't know if it is intuitive for others, but what about `it_mkLambda_or_dep_LetIn`?

**Comment by mattam82** [comment]
I can move it to Termops and rename it: its effect is to remove unused/non-dependent let-ins, so it_mkLambda_or_dep_LetIn sounds exactly right.

**Comment by maximedenes** [comment]
Sorry, I'm very tired again, so slow. Why do we `subst1 mkProp t`?

**Comment by mattam82** [comment]
As the term is not used in t we can substitute it by anything. We could use the body of the let instead I suppose, as there's no relocation involved.

**Comment by gares** [comment]
I think there is no point in traversing the term twice right?  Just do the subst1 (I assume subst1 relocates the substituted only if it occurs)

**Comment by gares** [comment]
LGTM

**Pull request merged**